File: packages/modules/blob/android/src/main/java/com/lynxpo/blob/BlobModule.java
// Copyright 2026 The Lynxpo Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynxpo.blob;

import android.content.Context;
import android.util.Base64;
import com.lynx.jsbridge.LynxMethod;
import com.lynx.jsbridge.LynxModule;
import com.lynx.jsbridge.LynxNativeModule;
import com.lynxpo.blob.generated.BlobModuleSpec;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Android counterpart of the iOS {@code BlobModule}. Exposes a real in-memory
 * binary large-object store to JS via {@code NativeModules.BlobModule},
 * faithfully porting the native surface of Expo's {@code expo-blob} (latest)
 * module. Method names MUST match the iOS {@code methodLookup} keys and the
 * shared {@code @lynxpo/mods-blob} accessors. Bytes are base64-encoded across
 * the bridge (Lynx supported types do not include raw byte arrays). The store
 * is process-scoped, exactly like Expo's native blob manager.
 */
@LynxNativeModule(name = "BlobModule")
public class BlobModule extends BlobModuleSpec {

  // blobId -> bytes
  private final Map<String, byte[]> store = new ConcurrentHashMap<>();
  // blobId -> type
  private final Map<String, String> types = new ConcurrentHashMap<>();

  public BlobModule(Context context) {
    super(context);
  }

  private static byte[] decode(String payload) {
    if (payload == null || payload.isEmpty()) return new byte[0];
    try {
      return Base64.decode(payload, Base64.NO_WRAP);
    } catch (IllegalArgumentException e) {
      return new byte[0];
    }
  }

  @LynxMethod
  public String create(String payload, String type) {
    try {
      byte[] data = decode(payload);
      String blobId = UUID.randomUUID().toString();
      store.put(blobId, data);
      types.put(blobId, type == null ? "" : type);
      return blobId;
    } catch (Exception e) {
      return null;
    }
  }

  @LynxMethod
  public double size(String blobId) {
    byte[] data = store.get(blobId);
    return data != null ? (double) data.length : 0.0;
  }

  @LynxMethod
  public String type(String blobId) {
    String t = types.get(blobId);
    return t != null ? t : "";
  }

  @LynxMethod
  public String slice(String blobId, double start, double end, String contentType) {
    try {
      byte[] data = store.get(blobId);
      if (data == null) return null;
      int len = data.length;
      int s = (int) start;
      int e = (int) end;
      if (s < 0) s = 0;
      if (e > len || e < 0) e = len;
      if (s > e) s = e;
      byte[] sliced = new byte[e - s];
      System.arraycopy(data, s, sliced, 0, e - s);
      String newId = UUID.randomUUID().toString();
      store.put(newId, sliced);
      types.put(newId, contentType == null ? "" : contentType);
      return newId;
    } catch (Exception ex) {
      return null;
    }
  }

  @LynxMethod
  public String bytes(String blobId) {
    try {
      byte[] data = store.get(blobId);
      if (data == null) return null;
      return Base64.encodeToString(data, Base64.NO_WRAP);
    } catch (Exception e) {
      return null;
    }
  }

  @LynxMethod
  public String text(String blobId) {
    try {
      byte[] data = store.get(blobId);
      if (data == null) return null;
      return new String(data, StandardCharsets.UTF_8);
    } catch (Exception e) {
      return null;
    }
  }

  @LynxMethod
  public String arrayBuffer(String blobId) {
    // ArrayBuffer of a blob is its raw bytes; base64-encoded across the bridge.
    return bytes(blobId);
  }
}

File: packages/modules/blob/ios/src/BlobModule.m
// Copyright 2026 The Lynxpo Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#import "BlobModule.h"

@implementation BlobModule {
  // blobId -> { data, type }
  NSMutableDictionary<NSString *, NSData *> *_store;
  NSMutableDictionary<NSString *, NSString *> *_types;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    _store = [NSMutableDictionary dictionary];
    _types = [NSMutableDictionary dictionary];
  }
  return self;
}

#pragma mark - Helpers

- (NSString *)nextId {
  return [[NSUUID UUID] UUIDString];
}

- (NSData *)decode:(NSString *)payload {
  // payload is a base64 string (UTF-8 text is base64-encoded by the JS facade).
  if (payload.length == 0) return [NSData data];
  return [[NSData alloc] initWithBase64EncodedString:payload
                                             options:NSDataBase64DecodingIgnoreUnknownCharacters]
             ?: [NSData data];
}

#pragma mark - Sync API

- (NSString *)create:(NSString *)payload type:(NSString *)type {
  @try {
    NSString *p = payload ?: @"";
    NSString *t = type ?: @"";
    NSData *data = [self decode:p];
    NSString *blobId = [self nextId];
    _store[blobId] = data;
    _types[blobId] = t;
    return blobId;
  } @catch (NSException *e) {
    return nil;
  }
}

- (double)size:(NSString *)blobId {
  NSData *data = _store[blobId];
  return data ? (double)data.length : 0.0;
}

- (NSString *)type:(NSString *)blobId {
  NSString *t = _types[blobId];
  return t ?: @"";
}

- (NSString *)slice:(NSString *)blobId
              start:(double)start
                end:(double)end
       contentType:(NSString *)contentType {
  @try {
    NSData *data = _store[blobId];
    if (!data) return nil;
    NSInteger len = (NSInteger)data.length;
    NSInteger s = (NSInteger)start;
    NSInteger e = (NSInteger)end;
    if (s < 0) s = 0;
    if (e > len || e < 0) e = len;
    if (s > e) s = e;
    NSRange range = NSMakeRange((NSUInteger)s, (NSUInteger)(e - s));
    NSData *sliced = [data subdataWithRange:range];
    NSString *blobId2 = [self nextId];
    _store[blobId2] = sliced;
    _types[blobId2] = contentType ?: @"";
    return blobId2;
  } @catch (NSException *e) {
    return nil;
  }
}

- (NSString *)bytes:(NSString *)blobId {
  @try {
    NSData *data = _store[blobId];
    if (!data) return nil;
    return [data base64EncodedStringWithOptions:0];
  } @catch (NSException *e) {
    return nil;
  }
}

- (NSString *)text:(NSString *)blobId {
  @try {
    NSData *data = _store[blobId];
    if (!data) return nil;
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  } @catch (NSException *e) {
    return nil;
  }
}

- (NSString *)arrayBuffer:(NSString *)blobId {
  // ArrayBuffer of a blob is its raw bytes; base64-encoded across the bridge.
  return [self bytes:blobId];
}

#pragma mark - LynxModule protocol

+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    @"create" : NSStringFromSelector(@selector(create:type:)),
    @"size" : NSStringFromSelector(@selector(size:)),
    @"type" : NSStringFromSelector(@selector(type:)),
    @"slice" : NSStringFromSelector(@selector(slice:start:end:contentType:)),
    @"bytes" : NSStringFromSelector(@selector(bytes:)),
    @"text" : NSStringFromSelector(@selector(text:)),
    @"arrayBuffer" : NSStringFromSelector(@selector(arrayBuffer:)),
  };
}

@end
